File: LAB-2/PYTHON/Left-Recursion.py

def Left_Recursion() :
    p = []
    temp = []
    result = []
    a = []
    b = []
    n = int(input("Enter no. of Grammars that you want :== "))
    print(" <<== Enter the Productions ==>> ")
    for i in range(0, n):
        p.append(input())
    for i in range(0, len(p)):

        t = []
        for j in range(0, len(p[i])):

            if (p[i][j] != "|"):
                t.append(p[i][j])
            else:
                temp.append(t)
                t = []
            if (j == len(p[i]) - 1):
                temp.append(t)

        for x in range(0, len(temp)):
            if (x == 0):
                if (temp[x][0] == temp[x][3]):
                    a.append(temp[x][4:len(temp[x]) + 1])

                else:
                    result.append(p[i])
                    break
            else:
                if (temp[x][0] == temp[0][0]):
                    a.append(temp[x][1:len(temp[x]) + 1])
                else:
                    b.append(temp[x])

        for i in range(0, len(a)):
            s = ""
            for j in range(0, len(a[i])):
                s = s + a[i][j]

            result.append(temp[0][0] + "'" + " -> " + s + temp[0][0] + "'" + " | E")
        for i in range(0, len(b)):
            s = ""
            for j in range(0, len(b[i])):
                s = s + b[i][j]

            result.append(temp[0][0] + " -> " + s + temp[0][0] + "'")
        temp = []
        t = []
        a = []
        b = []
    return result

if __name__ == "__main__" :
    production =Left_Recursion()
    print("\n <<=== Grammar ===>>\n")
    for result in production :
        print(result)

File: LAB-2/C++/Left-recursion.cpp
#include<iostream>
#include<vector>
#include<boost>
using namespace std;

void Left_Recursion(){
  int n;
  string x;
  std::vector<string> v,r,result;
  std::cout << " Enter the number of productions := ";
  std::cin >> n;
  std::cout << " <<==: Enter the Productions :==>> " << endl;
  for (int i=0;i<n;i++) {
    std::cin >> x;
    v.push_back(x);
  }
  for (int i=0;i<n;i++) {
    if (v[i].find("|")!= -1) {
      boost::split(r, v[i], boost::is_any_of("->"));

    }else{
      result.push_back(v[i]);
    }
  }

}

int main(){
  Left_Recursion();
  return 0;
}
